# Digital Lending Portfolio Intelligence & Risk-Adjusted Revenue Analytics Framework

### 1. Executive Summary
This project simulates how digital lending platfrms and retail banks monitor portfolio growth, credit risk, and profitability using loan-level data.

Using historical lending data, we analyze:

* Portfolio expansion trends
* Default rate dynamics  
* Revenue generation patterns  
* Risk-adjusted profitability  
* Segment-level performance

The objective is to evaluate whether rapid growth improved financial performance or introduced disproportionate credit risk.

### 2. Business Problem
Digital lenders and BNPL platforms often pursue aggresive growth. However, rapid loan issuance may increase creditrisk and reduce risk-adjusted profitability.

Key Question: 

Did portfolio expansion improve sustainable revenue, or did rising erode profitability?

### 3. Project Objectives

1. Analyze portfolio growth trends
2. Evaluate default rate acceleration during growth periods
3. Measure total interest income vs credit losses
4. Compute risk-adjusted revenue metrics
5. Identify high-risk vs high-profit segments
6. Design an analytics framework usable by banks and fintech lenders

### Why Python?

* Data cleaning and tranformation
* Handling large datasets efficiently
* Exploratory data analysis
* Aggregation and metric computation

It allows flexible financial metic calculations before loading structured outputs into PostgreSQL.

## Main questions 
1. Which borrower segments are most likely to default?
2. Which loan purposes generate the most losses?
3. Do income levels affect repayment?
4. Does loan grade predict default risk?
5. Does debt burden increase default risk?
6. Does interest rate reflect borrower risk?

## 1. Data Understanding and reduction of dataset

In [5]:
import pandas as pd
df_header = pd.read_csv("lending-club-loan-data-csv/loan.csv", nrows=0)
print(df_header.columns.tolist())

['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate', 'installment', 'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership', 'annual_inc', 'verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'url', 'desc', 'purpose', 'title', 'zip_code', 'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line', 'inq_last_6mths', 'mths_since_last_delinq', 'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'initial_list_status', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d', 'collections_12_mths_ex_med', 'mths_since_last_major_derog', 'policy_code', 'application_type', 'annual_inc_joint', 'dti_joint', 'verification_status_joint', 'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m', 'open_act_il', 'open_il_12m', 'open_i

**Note**:-Now we will use only those columns which are necessary related to our revenue analysis 

In [7]:
columns_needed = [
    'loan_amnt',
    'term',
    'int_rate',
    'installment',
    'issue_d',
    'loan_status',
    'total_pymnt',
    'total_rec_prncp',
    'total_rec_int',
    'total_rec_late_fee',
    'recoveries',
    'dti',
    'grade',
    'sub_grade',
    'annual_inc',
    'delinq_2yrs',
    'inq_last_6mths',
    'pub_rec_bankruptcies',
    'addr_state',
    'purpose',
    'home_ownership'
]

In [8]:
# we are doing this to know at which range of years the issued loans are high so we assume only that range which will reduce or unnecessary data
import pandas as pd

df_year = pd.read_csv(
    "lending-club-loan-data-csv/loan.csv",
    usecols=['issue_d'],
    parse_dates=['issue_d']
)

df_year['year'] = df_year['issue_d'].dt.year

year_counts = df_year['year'].value_counts().sort_index()

print(year_counts)

C:\Users\ayush\AppData\Local\Temp\ipykernel_7076\2070392532.py:4: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df_year = pd.read_csv(


year
2007       603
2008      2393
2009      5281
2010     12537
2011     21721
2012     53367
2013    134814
2014    235629
2015    421095
2016    434407
2017    443579
2018    495242
Name: count, dtype: int64


In [9]:
growth = year_counts.pct_change() * 100
print(growth)

year
2007           NaN
2008    296.849088
2009    120.685332
2010    137.398220
2011     73.255165
2012    145.693108
2013    152.616786
2014     74.780809
2015     78.711025
2016      3.161282
2017      2.111384
2018     11.646854
Name: count, dtype: float64


Result Till Now :- 
Data Scope
Time Period Selected: 2013–2017

Reason for Selection:

• Significant portfolio growth observed
• Clear acceleration in loan issuance
• Visible increase in default rates
• Sufficient data volume for trend analysis

In [11]:
file_path = "lending-club-loan-data-csv/loan.csv"

chunk_list = [] # it is going to store data in parts because data is big if we load it whole it takes time so we are extracting it in part chunk size =100000 it extracts every 1L data in list

for chunk in pd.read_csv(file_path, 
                         usecols=columns_needed, 
                         chunksize=100000):
    
    chunk['issue_d'] = pd.to_datetime(chunk['issue_d'],format='%b-%Y', errors='coerce')
    chunk['year'] = chunk['issue_d'].dt.year
    
    chunk = chunk[(chunk['year'] >= 2013) & (chunk['year'] <= 2017)]
    
    chunk_list.append(chunk)

df = pd.concat(chunk_list) # it converts and joins all parts in one dataframe in chunklist

print("Final Shape:", df.shape)
df.head()

Final Shape: (1669524, 22)


,loan_amnt,term,int_rate,installment,grade,sub_grade,home_ownership,annual_inc,issue_d,loan_status,...,dti,delinq_2yrs,inq_last_6mths,total_pymnt,total_rec_prncp,total_rec_int,total_rec_late_fee,recoveries,pub_rec_bankruptcies,year
495242,8000,36 months,12.79,268.75,C,C1,MORTGAGE,95000.0,2016-09-01,Current,...,10.30,0.0,1.0,7521.680000,5936.34,1585.34,0.0,0.00,1.0,2016
495243,20000,60 months,17.99,507.76,D,D2,MORTGAGE,66000.0,2016-09-01,Charged Off,...,12.62,2.0,0.0,7366.250000,2711.49,3341.65,0.0,1313.11,1.0,2016
495244,10000,36 months,8.59,316.10,A,A5,MORTGAGE,137500.0,2016-09-01,Fully Paid,...,13.42,0.0,1.0,10962.479076,10000.00,962.48,0.0,0.00,1.0,2016
495245,32200,60 months,21.49,880.02,D,D5,MORTGAGE,65000.0,2016-09-01,Fully Paid,...,11.71,0.0,1.0,35770.812746,32200.00,3570.81,0.0,0.00,1.0,2016
495246,2600,36 months,8.99,82.67,B,B1,RENT,35000.0,2016-09-01,Fully Paid,...,6.73,0.0,0.0,2662.590238,2600.00,62.59,0.0,0.00,0.0,2016


In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1669524 entries, 495242 to 2260667
Data columns (total 22 columns):
 #   Column                Non-Null Count    Dtype         
---  ------                --------------    -----         
 0   loan_amnt             1669524 non-null  int64         
 1   term                  1669524 non-null  object        
 2   int_rate              1669524 non-null  float64       
 3   installment           1669524 non-null  float64       
 4   grade                 1669524 non-null  object        
 5   sub_grade             1669524 non-null  object        
 6   home_ownership        1669524 non-null  object        
 7   annual_inc            1669524 non-null  float64       
 8   issue_d               1669524 non-null  datetime64[ns]
 9   loan_status           1669524 non-null  object        
 10  purpose               1669524 non-null  object        
 11  addr_state            1669524 non-null  object        
 12  dti                   1668945 non-null  fl

## 2. Feature Engineering & default rate calculation in each year

In [14]:
# Create default flag
df['is_default'] = df['loan_status'].isin(['Charged Off','Default']).astype(int)##assign 1 for loan status is charged off or default and 0 for other

# Remove current loans (optional but recommended for cleaner analysis)
df_analysis = df[df['loan_status'] != 'Current']

# Calculate yearly metrics
yearly_summary = df_analysis.groupby('year').agg(
    total_loans=('loan_status','count'),
    defaults=('is_default','sum')
)

yearly_summary['default_rate_%'] = (
    yearly_summary['defaults'] / yearly_summary['total_loans'] * 100
)

print(yearly_summary)

      total_loans  defaults  default_rate_%
year                                       
2013       134805     21022       15.594377
2014       222136     41056       18.482371
2015       375902     75275       20.025166
2016       282489     66679       23.604105
2017       171815     36389       21.179175


## Growth vs Risk Relationship

To evaluate whether aggressive expansion impacted portfolio quality, 
we compare year-over-year loan growth with changes in default rate.

This helps determine if rapid scaling led to deterioration in credit performance.

In [16]:
df['loan_status'].value_counts()

loan_status
Fully Paid            922873
Current               482377
Charged Off           240399
Late (31-120 days)     15388
In Grace Period         6051
Late (16-30 days)       2414
Default                   22
Name: count, dtype: int64

In [17]:
yearly_revenue = df_analysis.groupby('year').agg(
    total_interest=('total_rec_int','sum'),
    avg_interest_rate=('int_rate','mean')
)

print(yearly_revenue)

      total_interest  avg_interest_rate
year                                   
2013    4.633174e+08          14.531657
2014    6.585994e+08          13.649097
2015    9.288180e+08          12.394545
2016    6.029656e+08          13.237207
2017    2.752813e+08          14.084750


## Analysis Completed So Far

1. Year-wise Loan Growth
2. Year-wise Default Counts
3. Default Rate (%)
4. Growth vs Risk Relationship

Initial Insight:

Default rates increased during aggressive expansion 
(2013–2016), suggesting potential underwriting dilution.

**Business Interpretation – Revenue Trend**

Total interest income increased significantly from 2013 to 2015, 
peaking in 2015 as loan issuance expanded aggressively.

However, average interest rates declined during this expansion phase,
suggesting competitive pricing or relaxed underwriting standards.

Post-2015, total interest income declines sharply, likely due to:
• Portfolio seasoning effects
• Increased charge-offs
• Reduced net loan growth
• Incomplete loan maturity cycles

This indicates that revenue growth alone does not guarantee sustained profitability.

## Net Credit loss yearly
Outstanding Principal =
loan_amnt – total_rec_prncp

Then:

Net Credit Loss =
Outstanding Principal – recoveries
(Profitability analysis starts)

In [21]:
## Creating loss Column
df['outstanding_principal'] = df['loan_amnt'] - df['total_rec_prncp']

df['net_credit_loss'] = 0.0

df.loc[df['loan_status'] == 'Charged Off', 'net_credit_loss'] = \
    df['outstanding_principal'] - df['recoveries']

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1669524 entries, 495242 to 2260667
Data columns (total 25 columns):
 #   Column                 Non-Null Count    Dtype         
---  ------                 --------------    -----         
 0   loan_amnt              1669524 non-null  int64         
 1   term                   1669524 non-null  object        
 2   int_rate               1669524 non-null  float64       
 3   installment            1669524 non-null  float64       
 4   grade                  1669524 non-null  object        
 5   sub_grade              1669524 non-null  object        
 6   home_ownership         1669524 non-null  object        
 7   annual_inc             1669524 non-null  float64       
 8   issue_d                1669524 non-null  datetime64[ns]
 9   loan_status            1669524 non-null  object        
 10  purpose                1669524 non-null  object        
 11  addr_state             1669524 non-null  object        
 12  dti                    16689

In [23]:
## Aggregrate yearly
yearly_loss = df.groupby('year').agg(
    total_net_credit_loss=('net_credit_loss','sum'),
    total_recoveries=('recoveries','sum')
)

In [24]:
## merge with revenue table
final_financials = yearly_revenue.merge(
    yearly_loss,
    on='year',
    how='left'
)

final_financials['risk_adjusted_revenue'] = \
    final_financials['total_interest'] - final_financials['total_net_credit_loss']

final_financials

,total_interest,avg_interest_rate,total_net_credit_loss,total_recoveries,risk_adjusted_revenue
year,,,,,
2013,4.633174e+08,14.531657,1.827488e+08,2.764270e+07,2.805686e+08
2014,6.585994e+08,13.649097,3.627435e+08,5.509621e+07,2.958559e+08
2015,9.288180e+08,12.394545,7.188965e+08,9.579310e+07,2.099215e+08
2016,6.029656e+08,13.237207,6.825179e+08,7.814884e+07,-7.955230e+07
2017,2.752813e+08,14.084750,4.482860e+08,3.544736e+07,-1.730048e+08


### Business Interpretation – Risk-Adjusted Revenue

While total interest income peaked during high-growth years,
net credit losses also increased significantly.

Risk-adjusted revenue reveals whether portfolio expansion
improved true profitability or masked deteriorating credit qu

### Executive Insight – Risk-Adjusted Revenue Trend

Although total interest income increased significantly during 
the 2013–2015 expansion phase, net credit losses grew at a 
faster pace.

By 2016, risk-adjusted revenue turned negative, indicating 
that aggressive growth eroded portfolio profitability.

This suggests that loan expansion during high-growth years 
introduced disproportionate credit risk, leading to delayed 
economic impact in subsequent years.ality.

In [26]:
#Customer Risk Segmentation
def risk_segment(x):
    if x < 0.1:
        return "Low Risk"
    elif x < 0.2:
        return "Medium Risk"
    else:
        return "High Risk"

df['risk_segment'] = df['loan_status'].apply(lambda x: 1 if x == "Charged Off" else 0)

In [27]:
df['risk_segment'] = pd.qcut(df['int_rate'], 3, labels=["Low Risk","Medium Risk","High Risk"])

In [28]:
df['risk_adjusted_revenue'] = df['total_rec_int'] - df['net_credit_loss']

In [29]:
df.groupby('risk_segment')['int_rate'].mean()

C:\Users\ayush\AppData\Local\Temp\ipykernel_7076\3562418754.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('risk_segment')['int_rate'].mean()


risk_segment
Low Risk        8.526427
Medium Risk    12.851496
High Risk      18.598732
Name: int_rate, dtype: float64

In [30]:
df.groupby('risk_segment')['loan_status'].value_counts(normalize=True)

C:\Users\ayush\AppData\Local\Temp\ipykernel_7076\3384941680.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('risk_segment')['loan_status'].value_counts(normalize=True)


risk_segment  loan_status       
Low Risk      Fully Paid            0.617841
              Current               0.308883
              Charged Off           0.065420
              Late (31-120 days)    0.005056
              In Grace Period       0.002020
              Late (16-30 days)     0.000775
              Default               0.000005
Medium Risk   Fully Paid            0.554074
              Current               0.294454
              Charged Off           0.137152
              Late (31-120 days)    0.009112
              In Grace Period       0.003750
              Late (16-30 days)     0.001445
              Default               0.000013
High Risk     Fully Paid            0.481317
              Current               0.262108
              Charged Off           0.235339
              Late (31-120 days)    0.013808
              In Grace Period       0.005236
              Late (16-30 days)     0.002170
              Default               0.000022
Name: proportion, dtyp

In [31]:
df.groupby('risk_segment')['risk_adjusted_revenue'].mean()

C:\Users\ayush\AppData\Local\Temp\ipykernel_7076\2974792552.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('risk_segment')['risk_adjusted_revenue'].mean()


risk_segment
Low Risk       1085.228668
Medium Risk    1351.976558
High Risk      1571.022939
Name: risk_adjusted_revenue, dtype: float64

In [32]:
df

,loan_amnt,term,int_rate,installment,grade,sub_grade,home_ownership,annual_inc,issue_d,loan_status,...,total_rec_int,total_rec_late_fee,recoveries,pub_rec_bankruptcies,year,is_default,outstanding_principal,net_credit_loss,risk_segment,risk_adjusted_revenue
495242,8000,36 months,12.79,268.75,C,C1,MORTGAGE,95000.0,2016-09-01,Current,...,1585.34,0.0,0.00,1.0,2016,0,2063.66,0.0,Medium Risk,1585.34
495243,20000,60 months,17.99,507.76,D,D2,MORTGAGE,66000.0,2016-09-01,Charged Off,...,3341.65,0.0,1313.11,1.0,2016,1,17288.51,15975.4,High Risk,-12633.75
495244,10000,36 months,8.59,316.10,A,A5,MORTGAGE,137500.0,2016-09-01,Fully Paid,...,962.48,0.0,0.00,1.0,2016,0,0.00,0.0,Low Risk,962.48
495245,32200,60 months,21.49,880.02,D,D5,MORTGAGE,65000.0,2016-09-01,Fully Paid,...,3570.81,0.0,0.00,1.0,2016,0,0.00,0.0,High Risk,3570.81
495246,2600,36 months,8.99,82.67,B,B1,RENT,35000.0,2016-09-01,Fully Paid,...,62.59,0.0,0.00,0.0,2016,0,0.00,0.0,Low Risk,62.59
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2260663,12000,60 months,14.08,279.72,C,C3,MORTGAGE,58000.0,2017-10-01,Current,...,2048.16,0.0,0.00,0.0,2017,0,8687.20,0.0,Medium Risk,2048.16
2260664,12000,60 months,25.82,358.01,E,E4,MORTGAGE,30000.0,2017-10-01,Fully Paid,...,2499.80,0.0,0.00,0.0,2017,0,0.00,0.0,High Risk,2499.80
2260665,10000,36 months,11.99,332.10,B,B5,OWN,64000.0,2017-10-01,Current,...,1300.21,0.0,0.00,0.0,2017,0,5993.27,0.0,Medium Risk,1300.21
2260666,12000,60 months,21.45,327.69,D,D5,RENT,60000.0,2017-10-01,Current,...,3131.98,0.0,0.00,0.0,2017,0,9924.69,0.0,High Risk,3131.98


# 3. Loan Purpose vs defaul risk

In [34]:
purpose_default = df.groupby('purpose').agg(
    total_loans=('loan_status','count'),
    defaults=('loan_status', lambda x: (x == 'Charged Off').sum())
)

purpose_default['default_rate_%'] = \
(purpose_default['defaults'] / purpose_default['total_loans']) * 100

purpose_default.sort_values('default_rate_%', ascending=False)

,total_loans,defaults,default_rate_%
purpose,,,
small_business,16728,3488,20.851267
renewable_energy,993,176,17.724068
moving,11260,1913,16.989343
house,7862,1230,15.644874
debt_consolidation,967648,149285,15.427614
wedding,610,91,14.918033
medical,19483,2898,14.874506
other,96944,13874,14.311355
vacation,11280,1496,13.262411


### Borrower Risk Analysis by Loan Purpose

To understand which borrower categories contribute most to credit losses, we analyzed the default rate across different loan purposes.

The default rate was calculated as:

Default Rate = (Number of Charged Off Loans / Total Loans) 

Key Insights:

* Small business loans exhibit the highest default rate(~20.8%), indicating significantly higher credit risk compared to other loan categories.
* Debt consolidation represents the largest share of loans on the platform (over 960k loans). Even though its default rate is moderate (~15.4%), the high loan volume makes it a major contributor to the total credit loss.
* Credit card refinancing loans demonstrate relatively lower risk (~12.2%), suggesting that borrowers consolidating credit card debt may have more stable repayment patterns.
* Car loans show the lowest default rate (~10.5%), likely due to the presence of underlying collateral.

Business implication:
FinTech lenders can use this insight to optimize risk-based pricing strategies. High-risk categories such as small business lending may require stricter underwriting policies or higher interest rates to compensate for elevated default risk.evated default risk.

In [36]:
purpose_loss = df.groupby('purpose').agg(
    total_loans=('loan_status','count'),
    defaults=('is_default','sum'),
    total_loss=('net_credit_loss','sum')
)

purpose_loss['default_rate_%'] = (
    purpose_loss['defaults'] / purpose_loss['total_loans']
) * 100

purpose_loss = purpose_loss.sort_values('total_loss', ascending=False)

purpose_loss

,total_loans,defaults,total_loss,default_rate_%
purpose,,,,
debt_consolidation,967648,149299,1.552044e+09,15.429061
credit_card,373427,45724,4.406304e+08,12.244428
home_improvement,111617,13868,1.453638e+08,12.424631
other,96944,13875,1.042205e+08,14.312387
major_purchase,35152,4626,4.724520e+07,13.159991
small_business,16728,3488,4.014006e+07,20.851267
medical,19483,2898,1.963430e+07,14.874506
house,7862,1230,1.291455e+07,15.644874
car,16518,1736,1.188965e+07,10.509747


### Credit Loss Contribution by Loan Purpose

While default rate indicates the probability of borrower failure, it does not fully capture the financial impact on the lending platform.

To better understand portfolio risk, we calculate the total credit loss generated by each loan purpose.

This helps identify which borrower segments contribute the most to overall financial losses.

Preliminary observations suggest that categories such as debt consolidation and credit card refinancing dominate total losses due to their extremely high loan volumes, even if their default rates are lower than smaller categories such as small business loans.

This highlights the importance of analyzing both default probability and portfolio concentration when evaluating credit risk exposure.

## 4. Income risk analysis

In [39]:
## Borrower financial Strength analysis
# Do low income borrowers creates more defaults?
# Create income segments
df['income_group'] = pd.cut(
    df['annual_inc'],
    bins=[0,40000,80000,120000,200000,1000000],
    labels=['<40k','40k-80k','80k-120k','120k-200k','200k+']
)

income_default = df.groupby('income_group').agg(
    total_loans=('loan_status','count'),
    defaults=('is_default','sum')
)

income_default['default_rate_%'] = (
    income_default['defaults'] / income_default['total_loans']
) * 100

income_default

C:\Users\ayush\AppData\Local\Temp\ipykernel_7076\1076175217.py:10: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  income_default = df.groupby('income_group').agg(


,total_loans,defaults,default_rate_%
income_group,,,
<40k,293932,50973,17.341766
40k-80k,814074,123634,15.187071
80k-120k,362878,45254,12.470858
120k-200k,159849,17040,10.660060
200k+,37890,3444,9.089470


### Income vs Default Risk Analysis

Borrower income is a key factor in credit risk assessment. To understand its impact, borrowers were segmented into income groups and the default rate was calculated for each segment.

The results show a clear relationship between income level and default risk. Borrowers earning less than $40k annually exhibit the highest default rate (~17.3%), indicating greater financial vulnerability and limited repayment capacity.

As income increases, the default rate steadily declines. Borrowers earning above $200k have the lowest default rate (~9.1%), suggesting stronger financial stability and improved ability to meet loan obligations.

Interestingly, the majority of loans are issued to borrowers in the $40k–$80k income range, making this segment the largest contributor to overall defaults in absolute terms. However, their default rate remains lower than the lowest income segment.

These findings highlight income as a critical variable in credit risk evaluation and support the use of income-based risk segmentation in lending decisions.

## 5. Loan grade Risk analysis

In [42]:
## Loan Grade vs Default Risk
grade_risk = df.groupby('grade').agg(
    total_loans=('loan_status','count'),
    defaults=('is_default','sum'),
    avg_interest_rate=('int_rate','mean')
)

grade_risk['default_rate_%'] = (
    grade_risk['defaults'] / grade_risk['total_loans']
) * 100

grade_risk.sort_values('avg_interest_rate')

,total_loans,defaults,avg_interest_rate,default_rate_%
grade,,,,
A,276766,11858,7.055875,4.284486
B,491296,45848,10.541855,9.332052
C,502588,77602,13.985304,15.440480
D,242039,54626,17.814887,22.569090
E,110102,32958,21.413410,29.934061
F,36009,13302,25.408851,36.940765
G,10724,4227,28.353715,39.416263


## Loan Grade VS Default Risk

Loan Grades represent the lender's internal credit risk assesment of borrowers. To evaluate whether this grading system accurately reflects borrower risk, we analyzed default rates and average interest rates across loan grades.

The result show a strong and consistent relationship between loan grade, interest rate, and default probability. Borrowers in Grade A exhibit the lowest default rate (approximately 4.3%) and are chargedthe lower interest rate (~7%). As the credit grade worsens, both the interest rate and the default probability increase significantly.

For example, Grade D oans carry an average interest rate of around 17.8% with a default rate exceeding 22%, while Grade G loan exhibit the highest default rate (39%) and the highest average interest rate (28%).

This pattern confirms that the platform employs a risk-based pricing model, whereborrowers assessed as higher riskare charged higher interest rates to comensate for potenial losses. The clear monotonic increase in default rates across grads suggests that the credit gradingsystem effectively differentiates borrower risk levels.

Additionally, the majority of loans are concentrated in the B and C grade segments, indicating that the lending platform primarily serves borrowers with moderate credit risk.

## 6. Debt To income ratio

In [45]:
## Debt-to-Income Ratio (DTI).
## DTI = monthly debt / monthly income

# Create DTI segments
df['dti_group'] = pd.cut(
    df['dti'],
    bins=[0,10,20,30,40,100],
    labels=['0-10','10-20','20-30','30-40','40+']
)

dti_risk = df.groupby('dti_group').agg(
    total_loans=('loan_status','count'),
    defaults=('is_default','sum')
)

dti_risk['default_rate_%'] = (
    dti_risk['defaults'] / dti_risk['total_loans']
) * 100

dti_risk

C:\Users\ayush\AppData\Local\Temp\ipykernel_7076\2715406155.py:11: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  dti_risk = df.groupby('dti_group').agg(


,total_loans,defaults,default_rate_%
dti_group,,,
0-10,283751,30679,10.811944
10-20,686867,88958,12.951270
20-30,521341,85787,16.455065
30-40,163873,33168,20.240064
40+,11455,1609,14.046268


In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1669524 entries, 495242 to 2260667
Data columns (total 29 columns):
 #   Column                 Non-Null Count    Dtype         
---  ------                 --------------    -----         
 0   loan_amnt              1669524 non-null  int64         
 1   term                   1669524 non-null  object        
 2   int_rate               1669524 non-null  float64       
 3   installment            1669524 non-null  float64       
 4   grade                  1669524 non-null  object        
 5   sub_grade              1669524 non-null  object        
 6   home_ownership         1669524 non-null  object        
 7   annual_inc             1669524 non-null  float64       
 8   issue_d                1669524 non-null  datetime64[ns]
 9   loan_status            1669524 non-null  object        
 10  purpose                1669524 non-null  object        
 11  addr_state             1669524 non-null  object        
 12  dti                    16689

In [47]:
df['dti']

495242     10.30
495243     12.62
495244     13.42
495245     11.71
495246      6.73
           ...  
2260663    20.88
2260664    19.28
2260665    12.96
2260666    30.82
2260667    18.40
Name: dti, Length: 1669524, dtype: float64

In [48]:
df

,loan_amnt,term,int_rate,installment,grade,sub_grade,home_ownership,annual_inc,issue_d,loan_status,...,recoveries,pub_rec_bankruptcies,year,is_default,outstanding_principal,net_credit_loss,risk_segment,risk_adjusted_revenue,income_group,dti_group
495242,8000,36 months,12.79,268.75,C,C1,MORTGAGE,95000.0,2016-09-01,Current,...,0.00,1.0,2016,0,2063.66,0.0,Medium Risk,1585.34,80k-120k,10-20
495243,20000,60 months,17.99,507.76,D,D2,MORTGAGE,66000.0,2016-09-01,Charged Off,...,1313.11,1.0,2016,1,17288.51,15975.4,High Risk,-12633.75,40k-80k,10-20
495244,10000,36 months,8.59,316.10,A,A5,MORTGAGE,137500.0,2016-09-01,Fully Paid,...,0.00,1.0,2016,0,0.00,0.0,Low Risk,962.48,120k-200k,10-20
495245,32200,60 months,21.49,880.02,D,D5,MORTGAGE,65000.0,2016-09-01,Fully Paid,...,0.00,1.0,2016,0,0.00,0.0,High Risk,3570.81,40k-80k,10-20
495246,2600,36 months,8.99,82.67,B,B1,RENT,35000.0,2016-09-01,Fully Paid,...,0.00,0.0,2016,0,0.00,0.0,Low Risk,62.59,<40k,0-10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2260663,12000,60 months,14.08,279.72,C,C3,MORTGAGE,58000.0,2017-10-01,Current,...,0.00,0.0,2017,0,8687.20,0.0,Medium Risk,2048.16,40k-80k,20-30
2260664,12000,60 months,25.82,358.01,E,E4,MORTGAGE,30000.0,2017-10-01,Fully Paid,...,0.00,0.0,2017,0,0.00,0.0,High Risk,2499.80,<40k,10-20
2260665,10000,36 months,11.99,332.10,B,B5,OWN,64000.0,2017-10-01,Current,...,0.00,0.0,2017,0,5993.27,0.0,Medium Risk,1300.21,40k-80k,10-20
2260666,12000,60 months,21.45,327.69,D,D5,RENT,60000.0,2017-10-01,Current,...,0.00,0.0,2017,0,9924.69,0.0,High Risk,3131.98,40k-80k,30-40


Key findings:

1️. Low income borrowers have higher default rates

2️. Higher DTI increases risk

3️. Loan grades correctly predict risk

4️. Higher interest rates compensate for higher risk

5️. Medium-risk borrowers dominate the portfolio

In [50]:
df.isna().sum()

loan_amnt                   0
term                        0
int_rate                    0
installment                 0
grade                       0
sub_grade                   0
home_ownership              0
annual_inc                  0
issue_d                     0
loan_status                 0
purpose                     0
addr_state                  0
dti                       579
delinq_2yrs                 0
inq_last_6mths              1
total_pymnt                 0
total_rec_prncp             0
total_rec_int               0
total_rec_late_fee          0
recoveries                  0
pub_rec_bankruptcies        0
year                        0
is_default                  0
outstanding_principal       0
net_credit_loss             0
risk_segment                0
risk_adjusted_revenue       0
income_group              901
dti_group                2237
dtype: int64

In [51]:
"""df.to_csv("cleaned_loan_dataset.csv", index=False) ## to save clean dataset"""

'df.to_csv("cleaned_loan_dataset.csv", index=False) ## to save clean dataset'

In [52]:
"""pip install sqlalchemy psycopg2 pandas"""

'pip install sqlalchemy psycopg2 pandas'

In [53]:
"""import pandas as pd
from sqlalchemy import create_engine

engine = create_engine("postgresql://postgres:%40yush170@localhost:5432/loan_risk_db")

file_path = r"C:\Users\ayush\Downloads\cleaned_loan_dataset.csv"

chunksize = 50000

for chunk in pd.read_csv(file_path, chunksize=chunksize):
    chunk.to_sql("loan_data_clean", engine, if_exists="append", index=False)

print("Data upload completed!")"""

SyntaxError: (unicode error) 'unicodeescape' codec can't decode bytes in position 162-163: truncated \UXXXXXXXX escape (1398586216.py, line 1)

In [54]:
df.sample(5000).to_csv("sample_loan_data.csv", index=False)

In [56]:
pd.read_csv('sample_loan_data.csv')

,loan_amnt,term,int_rate,installment,grade,sub_grade,home_ownership,annual_inc,issue_d,loan_status,...,recoveries,pub_rec_bankruptcies,year,is_default,outstanding_principal,net_credit_loss,risk_segment,risk_adjusted_revenue,income_group,dti_group
0,8000,36 months,11.99,265.68,C,C1,RENT,49000.0,2016-03-01,Fully Paid,...,0.0,0.0,2016,0,0.00,0.0,Medium Risk,1135.47,40k-80k,10-20
1,30575,36 months,11.99,1015.39,B,B5,MORTGAGE,150000.0,2017-08-01,Current,...,0.0,0.0,2017,0,16651.60,0.0,Medium Risk,4333.25,120k-200k,20-30
2,35000,36 months,13.98,1195.88,C,C3,MORTGAGE,135000.0,2014-08-01,Fully Paid,...,0.0,0.0,2014,0,0.00,0.0,Medium Risk,4369.27,120k-200k,10-20
3,3000,36 months,12.74,100.71,C,C1,MORTGAGE,30000.0,2016-10-01,Fully Paid,...,0.0,0.0,2016,0,0.00,0.0,Medium Risk,144.92,<40k,20-30
4,35000,36 months,12.99,1179.12,C,C2,MORTGAGE,200000.0,2016-05-01,Current,...,0.0,0.0,2016,0,3462.12,0.0,Medium Risk,7347.82,120k-200k,10-20
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4995,14400,60 months,12.29,322.44,C,C1,MORTGAGE,59000.0,2015-10-01,Fully Paid,...,0.0,0.0,2015,0,0.00,0.0,Medium Risk,2577.45,40k-80k,10-20
4996,10000,36 months,10.99,327.34,B,B2,OWN,130000.0,2014-03-01,Fully Paid,...,0.0,1.0,2014,0,0.00,0.0,Low Risk,193.72,120k-200k,0-10
4997,6000,36 months,12.69,201.27,C,C2,RENT,22000.0,2015-05-01,Fully Paid,...,0.0,0.0,2015,0,0.00,0.0,Medium Risk,1241.45,<40k,10-20
4998,5750,36 months,19.48,212.18,E,E2,MORTGAGE,60000.0,2016-01-01,Fully Paid,...,0.0,0.0,2016,0,0.00,0.0,High Risk,1849.25,40k-80k,10-20
